### PR TITLE
Fix viewer.reset() throws TypeError

### LINKED ISSUE #4867: `viewer.reset()` throws `TypeError`
## 🐛 Bug
Calling `viewer.reset()` via the napari console results in a `TypeError`.

## To Reproduce

1. Open an image e.g. XYZ tif.
2. Open iPython console.
3. Run `viewer.reset()`.

```
viewer.reset()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [1], in <cell line: 1>()
----> 1 viewer.reset()

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/events/evented_model.py:253, in EventedModel.reset(self)
    251 for name, value in self._defaults.items():
    252     if isinstance(value, EventedModel):
--> 253         getattr(self, name).reset()
    254     elif (
    255         self.__config__.allow_mutation
    256         and self.__fields__[name].field_info.allow_mutation
    257     ):
    258         setattr(self, name, value)

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/events/evented_model.py:253, in EventedModel.reset(self)
    251 for name, value in self._defaults.items():
    252     if isinstance(value, EventedModel):
--> 253         getattr(self, name).reset()
    254     elif (
    255         self.__config__.allow_mutation
    256         and self.__fields__[name].field_info.allow_mutation
    257     ):
    258         setattr(self, name, value)

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/events/evented_model.py:258, in EventedModel.reset(self)
    253     getattr(self, name).reset()
    254 elif (
    255     self.__config__.allow_mutation
    256     and self.__fields__[name].field_info.allow_mutation
    257 ):
--> 258     setattr(self, name, value)

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/events/evented_model.py:234, in EventedModel.__setattr__(self, name, value)
    232 are_equal = self.__eq_operators__.get(name, operator.eq)
    233 if not are_equal(after, before):
--> 234     getattr(self.events, name)(value=after)  # emit event
    236     # emit events for any dependent computed property setters as well
    237     for dep in self.__field_dependents__.get(name, {}):

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/events/event.py:757, in EventEmitter.__call__(self, *args, **kwargs)
    754     self._block_counter.update([cb])
    755     continue
--> 757 self._invoke_callback(cb, event if pass_event else None)
    758 if event.blocked:
    759     break

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/events/event.py:794, in EventEmitter._invoke_callback(self, cb, event)
    792     self.disconnect(cb)
    793     return
--> 794 _handle_exception(
    795     self.ignore_callback_errors,
    796     self.print_callback_errors,
    797     self,
    798     cb_event=(cb, event),
    799 )

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/events/event.py:781, in EventEmitter._invoke_callback(self, cb, event)
    779 try:
    780     if event is not None:
--> 781         cb(event)
    782     else:
    783         cb()

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/components/_interaction_box_mouse_bindings.py:192, in InteractionBoxMouseBindings._on_transform_change(self, event)
    184 """Gets called when the interaction box is transformed to update transform of the layer"""
    186 layer_dims_displayed = dims_displayed_world_to_layer(
    187     list(self._ref_viewer().dims.displayed),
    188     self._ref_viewer().dims.ndim,
    189     self._ref_viewer().layers.selection.active.ndim,
    190 )
--> 192 layer_affine_transform = event.value.compose(
    193     self._initial_transform_inverse
    194 )
    196 self._ref_viewer().layers.selection.active.affine = (
    197     self._ref_viewer().layers.selection.active.affine.replace_slice(
    198         layer_dims_displayed, layer_affine_transform
    199     )
    200 )

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/transforms/transforms.py:552, in Affine.compose(self, transform)
    550 """Return the composite of this transform and the provided one."""
    551 if not isinstance(transform, Affine):
--> 552     return super().compose(transform)
    553 affine_matrix = self.affine_matrix @ transform.affine_matrix
    554 return Affine(affine_matrix=affine_matrix)

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/transforms/transforms.py:60, in Transform.compose(self, transform)
     58 def compose(self, transform: 'Transform') -> 'Transform':
     59     """Return the composite of this transform and the provided one."""
---> 60     return TransformChain([self, transform])

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/transforms/transforms.py:116, in TransformChain.__init__(self, transforms)
    114 if transforms is None:
    115     transforms = []
--> 116 super().__init__(
    117     data=transforms,
    118     basetype=Transform,
    119     lookup={str: lambda x: x.name},
    120 )
    121 # The above super().__init__() will not call Transform.__init__().
    122 # For that to work every __init__() called using super() needs to
    123 # in turn call super().__init__(). So we call it explicitly here.
    124 Transform.__init__(self)

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/events/containers/_evented_list.py:101, in EventedList.__init__(self, data, basetype, lookup)
     98 else:
     99     # otherwise create a new one
    100     self.events = EmitterGroup(source=self, **_events)
--> 101 super().__init__(data, basetype=basetype, lookup=lookup)

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/events/containers/_typed.py:64, in TypedMutableSequence.__init__(self, data, basetype, lookup)
     60 self._basetypes = (
     61     basetype if isinstance(basetype, Sequence) else (basetype,)
     62 )
     63 self._lookup = lookup.copy()
---> 64 self.extend(data)

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/_collections_abc.py:983, in MutableSequence.extend(self, values)
    981     values = list(values)
    982 for v in values:
--> 983     self.append(v)

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/_collections_abc.py:962, in MutableSequence.append(self, value)
    960 def append(self, value):
    961     'S.append(value) -- append value to the end of the sequence'
--> 962     self.insert(len(self), value)

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/events/containers/_evented_list.py:182, in EventedList.insert(self, index, value)
    180 """Insert ``value`` before index."""
    181 self.events.inserting(index=index)
--> 182 super().insert(index, value)
    183 self.events.inserted(index=index, value=value)
    184 self._connect_child_emitters(value)

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/events/containers/_typed.py:102, in TypedMutableSequence.insert(self, index, value)
    101 def insert(self, index: int, value: _T):
--> 102     self._list.insert(index, self._type_check(value))

File ~/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/napari/utils/events/containers/_typed.py:160, in TypedMutableSequence._type_check(self, e)
    156 def _type_check(self, e: Any) -> _T:
    157     if self._basetypes and not any(
    158         isinstance(e, t) for t in self._basetypes
    159     ):
--> 160         raise TypeError(
    161             trans._(
    162                 'Cannot add object with type {dtype!r} to TypedList expecting type {basetypes!r}',
    163                 deferred=True,
    164                 dtype=type(e),
    165                 basetypes=self._basetypes,
    166             )
    167         )
    168     return e

TypeError: Cannot add object with type <class 'NoneType'> to TypedList expecting type (<class 'napari.utils.transforms.transforms.Transform'>,)
/home/tibuch/Apps/miniconda3/envs/psf-analysis/lib/python3.8/site-packages/ipykernel/kernelbase.py:757: RuntimeWarning: coroutine 'InProcessKernel._abort_queues' was never awaited!
```

## Expected behavior
Reset the viewer perspective.

## Environment

 - Please copy and paste the information at napari info option in help menubar here:
```
napari: 0.4.16
Platform: Linux-5.14.0-1038-oem-x86_64-with-glibc2.17
System: Ubuntu 22.04 LTS
Python: 3.8.13 (default, Mar 28 2022, 11:38:47) [GCC 7.5.0]
Qt: 5.15.2
PyQt5: 5.15.7
NumPy: 1.23.1
SciPy: 1.8.1
Dask: 2022.7.1
VisPy: 0.10.0

OpenGL:
- GL version: 4.6.0 NVIDIA 470.129.06
- MAX_TEXTURE_SIZE: 32768

Screens:
- screen 1: resolution 3440x1440, scale 1.0
- screen 2: resolution 1920x1080, scale 1.0

Plugins:
- bfio: 2.3.0
- console: 0.0.4
- napari-aicsimageio: 0.7.1
- napari-psf-analysis: 0.2.0
- napari-svg: 0.1.6
- ome-types: 0.3.0
- scikit-image: 0.4.16
```

 - Any other relevant information:

## Additional context
N/A

### PR DESCRIPTION
**Pull request by Czaki**
# Description

Fix raising error in napari Viewer on call `reset` method. 

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# References

close #4867

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: napari/_tests/test_viewer.py [hunk 1]
```diff
@@ -357,3 +357,16 @@ def test_current_viewer(make_napari_viewer):
 
     assert current_viewer() is viewer1
     assert current_viewer() is not viewer2
+
+
+def test_reset_empty(make_napari_viewer):
+    """Test that resetting an empty viewer doesn't crash"""
+    viewer = make_napari_viewer()
```
Comment (andy-sweet) [inline napari/_tests/test_viewer.py:367]: As `reset` is defined in `EventedModel` and the bug this reproduces should occur with `ViewerModel`, I think we should test that instead of the heavier full viewer.

Comment (andy-sweet) [reply napari/_tests/test_viewer.py:367]: Ah wait, sorry, maybe `ViewerModel` is not enough - I could *not* reproduce with it on `main`. Ignore if that's true.

File: napari/_tests/test_viewer.py [hunk 2]
```diff
@@ -357,3 +357,16 @@ def test_current_viewer(make_napari_viewer):
...
+    viewer.reset()
+
+
+def test_reset_non_empty(make_napari_viewer):
+    """Test that resetting an empty viewer doesn't crash"""
```
Comment (andy-sweet) [inline napari/_tests/test_viewer.py]:
```suggestion
    """Test that resetting a non-empty viewer doesn't crash"""
```

File: napari/_tests/test_viewer.py [hunk 3]
```diff
@@ -357,3 +357,16 @@ def test_current_viewer(make_napari_viewer):
 
     assert current_viewer() is viewer1
     assert current_viewer() is not viewer2
+
+
+def test_reset_empty(make_napari_viewer):
+    """Test that resetting an empty viewer doesn't crash"""
```
Comment (andy-sweet) [inline napari/_tests/test_viewer.py]: Optional: as this test doesn't assert anything and only checks that there is no error, a link to the original issue would be useful for context. I guess we can assume people can find it via blame, so if you think this isn't important, feel free to ignore.

File: napari/components/_interaction_box_mouse_bindings.py [hunk 4]
```diff
@@ -183,20 +183,23 @@ def _on_mode_change(self, event):
...
+        active_layer = viewer.layers.selection.active
+        if active_layer is None:
+            return
+
         layer_dims_displayed = dims_displayed_world_to_layer(
             list(self._ref_viewer().dims.displayed),
-            self._ref_viewer().dims.ndim,
-            self._ref_viewer().layers.selection.active.ndim,
+            viewer.dims.ndim,
+            active_layer.ndim if active_layer is not None else 2,
```
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py]:
```suggestion
            active_layer.ndim,
```

Given the early return above, I think this ternary statement will always evaluate to `active_layer.ndim`. That allows us to remove the magic number `2` here.

File: napari/utils/transforms/transforms.py [hunk 5]
```diff
@@ -55,9 +55,9 @@ def inverse(self) -> 'Transform':
                 trans._('Inverse function was not provided.', deferred=True)
             )
 
-    def compose(self, transform: 'Transform') -> 'Transform':
+    def compose(self, transform: Optional['Transform']) -> 'Transform':
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]: I don't love changing the signature of this method because other classes override it and also need to deal with `None` now. I'd prefer to make changes to `_interaction_box_mouse_bindings.py` to handle the case when the input is `None`, as I would expect an error if I compose `None` with an existing transform. What do you think?

File: napari/_tests/test_viewer.py (hunk 1 above)
Comment (Czaki) [reply napari/_tests/test_viewer.py:367]: I do not know. I could not reproduce the error using `ViewerModel`. I just want to fix this bug before release.

File: napari/utils/transforms/transforms.py (hunk 5 above)
Comment (Czaki) [reply napari/utils/transforms/transforms.py]: I implement simplest solution. If you better understand these files feel free to implement fixed version

File: napari/components/_interaction_box_mouse_bindings.py [hunk 6]
```diff
@@ -183,20 +183,23 @@ def _on_mode_change(self, event):
...
             list(self._ref_viewer().dims.displayed),
-            self._ref_viewer().dims.ndim,
-            self._ref_viewer().layers.selection.active.ndim,
+            viewer.dims.ndim,
+            active_layer.ndim,
         )
 
         layer_affine_transform = event.value.compose(
             self._initial_transform_inverse
         )
```
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py]: Instead of changing `Transform.compose` to handle `None` (which is an API change), we could just special case things here as a simpler quick fix.

```suggestion
        layer_affine_transform = event.value
        if self._initial_transform_inverse is not None:
            layer_affine_transform = layer_affine_transform.compose(self._initial_transform_inverse)
```

File: napari/utils/transforms/transforms.py [hunk 7]
```diff
@@ -55,9 +55,9 @@ def inverse(self) -> 'Transform':
...
         """Return the composite of this transform and the provided one."""
-        return TransformChain([self, transform])
+        return self if transform is None else TransformChain([self, transform])
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]: My suggestion is to revert the change here

```suggestion
        return TransformChain([self, transform])
```

and special case them in the interaction box code: https://github.com/napari/napari/pull/4868/files#r930376668

File: napari/_tests/test_viewer.py (hunk 1 above)
Comment (andy-sweet) [reply napari/_tests/test_viewer.py:367]: Agreed - I cannot reproduce it with `ViewerModel` either. There must be something about `Viewer` that's causing the issue. Not sure what that is, but I understand the changes here enough to approve.

File: napari/utils/transforms/transforms.py (hunk 5 above)
Comment (andy-sweet) [reply napari/utils/transforms/transforms.py]: I made a couple of line suggestions to show what I meant.

File: napari/components/_interaction_box_mouse_bindings.py (hunk 6 above)
Comment (Czaki) [inline napari/components/_interaction_box_mouse_bindings.py]:
```suggestion
        layer_affine_transform = event.value.compose(
            self._initial_transform_inverse
        ) if self._initial_transform_inverse is not None else event.value
```

Comment (Czaki) [reply napari/components/_interaction_box_mouse_bindings.py]: oh. I miss `layer_affine_transform = event.value` line and implement my own version.

File: napari/utils/transforms/transforms.py [hunk 8]
```diff
@@ -1,5 +1,5 @@
 from functools import cached_property
-from typing import Sequence
+from typing import Optional, Sequence
```
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]:
```suggestion
from typing import Sequence
```

File: napari/utils/transforms/transforms.py (hunk 5 above)
Comment (andy-sweet) [reply napari/utils/transforms/transforms.py]:
```suggestion
    def compose(self, transform: 'Transform') -> 'Transform':
```

**Review by andy-sweet — APPROVED** [review]
Looks good to me, except for a couple of other suggestions.

**Comment by andy-sweet** [comment]
Might be good for someone more familiar with the interaction box code to look at this, like @jojoelfe ? There might be a way to avoid the `None` value of the initial transform entirely.

**Comment by Carreau** [comment]
> Might be good for someone more familiar with the interaction box code to look at this, like @jojoelfe ? There might be a way to avoid the `None` value of the initial transform entirely.

I think we can set `_initial_transform(_inverse?)` in `napari/components/_interaction_box_mouse_bindings.py` to `Affine()` which IIUC should be the identity ?

**Comment by Czaki** [comment]
> I think we can set `_initial_transform(_inverse?)` in `napari/components/_interaction_box_mouse_bindings.py` to `Affine()` which IIUC should be the identity ?

I do not like this. It will create a chain containing identity functions that will lead to multiple obsolete calls. 

Of course, it could be done better, and we could add some checks that will prevent extending the chain by identity.

**Comment by Carreau** [comment]
So can we have a `class Identity(Transform)`, that special case identity in transforms themselves ? 

I'm mostly trying to find a solution to @andy-sweet comment, dealing with None here seem fine for now.

Maybe merge as-is and open a tracking issue to deal with it in transform.

**Comment by Carreau** [comment]
Ok, the rebase I did before the week-end is working, so merging and opening an issue to maybe have an identity transform.

**Comment by andy-sweet** [comment]
I think it's fine to merge this. Thanks for opening the issue @Carreau.
